plan/logical_plan_builder.go

```diff
@@ -1212,43 +1213,94 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) LogicalPlan {
 			pkCol = schema.Columns[schema.Len()-1]
 		}
 	}
+	p.SetSchema(schema)
+
 	needUnionScan := b.ctx.Txn() != nil && !b.ctx.Txn().IsReadOnly()
-	if b.needColHandle == 0 && !needUnionScan {
-		p.SetSchema(schema)
-		return p
+	if b.needColHandle != 0 || needUnionScan {
+		if pkCol == nil || needUnionScan {
+			idCol := &expression.Column{
+				FromID:   p.id,
+				DBName:   schemaName,
+				TblName:  tableInfo.Name,
+				ColName:  model.NewCIStr("_rowid"),
+				RetType:  types.NewFieldType(mysql.TypeLonglong),
+				Position: schema.Len(),
+				Index:    schema.Len(),
+				ID:       model.ExtraHandleID,
+			}
+			if needUnionScan {
+				p.unionScanSchema = expression.NewSchema(make([]*expression.Column, 0, len(tableInfo.Columns))...)
+				for _, col := range p.schema.Columns {
+					p.unionScanSchema.Append(col)
+				}
+				if b.needColHandle > 0 {
+					p.unionScanSchema.Columns = append(p.unionScanSchema.Columns, idCol)
+					p.unionScanSchema.TblID2Handle[tableInfo.ID] = []*expression.Column{idCol}
+				}
+			}
+			p.Columns = append(p.Columns, &model.ColumnInfo{
+				ID:   model.ExtraHandleID,
+				Name: model.NewCIStr("_rowid"),
+			})
+			p.schema.Append(idCol)
+			p.schema.TblID2Handle[tableInfo.ID] = []*expression.Column{idCol}
+		} else {
+			p.schema.TblID2Handle[tableInfo.ID] = []*expression.Column{pkCol}
+		}
 	}
-	if pkCol == nil || needUnionScan {
-		idCol := &expression.Column{
-			FromID:   p.id,
-			DBName:   schemaName,
-			TblName:  tableInfo.Name,
-			ColName:  model.NewCIStr("_rowid"),
-			RetType:  types.NewFieldType(mysql.TypeLonglong),
-			Position: schema.Len(),
-			Index:    schema.Len(),
-			ID:       model.ExtraHandleID,
-		}
-		if needUnionScan {
-			p.unionScanSchema = expression.NewSchema(make([]*expression.Column, 0, len(tableInfo.Columns))...)
-			for _, col := range schema.Columns {
-				p.unionScanSchema.Append(col)
+	return b.projectVirtualColumns(p, columns)
+}
+
+// projectVirtualColumns is only for DataSource. If some table has virtual generated columns,
+// we add a projection on the original DataSource, and calculate those columns in the projection
+// so that plans above it can reference generated columns by their name.
+func (b *planBuilder) projectVirtualColumns(ds *DataSource, columns []*table.Column) LogicalPlan {
+	if b.inUpdateStmt {
```

Comment:
If the DataSource is in `Update`, we don't need to calculate any generated columns, because we want to update them but not fetch back them. So we can skip this projection, for optimization.